launchpad/src/launchpad.rs

```diff
@@ -3,324 +3,340 @@
             "Wrong payment token used"
         );
-        require!(payment_amount == total_ticket_price, "Wrong amount sent");
-
-        let (first_ticket_id, last_ticket_id) = self.try_get_ticket_range(&caller)?;
-        let nr_tickets = last_ticket_id - first_ticket_id + 1;
-
-        require!(
-            nr_tickets >= nr_tickets_to_confirm,
-            "Trying to confirm too many tickets"
-        );
-
-        let current_generation = self.current_generation().get();
-        let winning_tickets = self.get_tickets_with_status(
-            &caller,
-            TicketStatus::Winning {
-                generation: current_generation,
-            },
-        );
-
         require!(
-            nr_tickets_to_confirm <= winning_tickets.len(),
-            "Trying to confirm too many tickets"
+            payment_amount == total_ticket_price,
+            "Wrong amount sent, must confirm all tickets"
         );
 
-        for winning_ticket_id in &winning_tickets[..nr_tickets_to_confirm] {
-            self.ticket_status(*winning_ticket_id)
-                .set(&TicketStatus::Confirmed {
-                    generation: current_generation,
-                });
-        }
-
-        self.total_confirmed_tickets()
-            .update(|confirmed| *confirmed += nr_tickets_to_confirm);
+        self.confirmed_all_tickets(&caller).set(&true);
 
         Ok(())
     }
 
-    #[endpoint(selectNewWinners)]
-    fn select_new_winners(&self) -> SCResult<BoxedBytes> {
-        self.require_stage(LaunchStage::SelectNewWinners)?;
+    #[only_owner]
+    #[endpoint(filterTickets)]
+    fn filter_tickets(&self) -> SCResult<BoxedBytes> {
+        self.require_winner_selection_period()?;
+        require!(!self.tickets_filtered().get(), "Tickets already filtered");
 
-        let (_, prev_last_winning_ticket_position) = self.winning_tickets_range().get();
-        let new_first_winning_ticket_position = prev_last_winning_ticket_position + 1;
+        let last_ticket_id = self.last_ticket_id().get();
+        let (mut first_ticket_id_in_batch, mut nr_removed) =
+            self.load_filter_tickets_operation()?;
 
-        let (mut current_ticket_position, winning_tickets) =
-            self.load_select_new_winners_operation(new_first_winning_ticket_position)?;
-        let confirmed_tickets = self.total_confirmed_tickets().get();
-        let remaining_tickets = winning_tickets - confirmed_tickets;
+        let run_result = self.run_while_it_has_gas(|| {
+            let (address, nr_tickets_in_batch) = self.ticket_batch(first_ticket_id_in_batch).get();
+
+            if !self.is_user_eligible(&address) {
+                nr_removed += nr_tickets_in_batch;
+
+                self.ticket_range_for_address(&address).clear();
+                self.ticket_batch(first_ticket_id_in_batch).clear();
+            } else if nr_removed > 0 {
+                let new_first_id = first_ticket_id_in_batch - nr_removed;
```

Comment:
this will be changed as well when we consider that a user can select how many tickets he/she wants to confirm